zajecia/src/optionalAndLambda/Main.java

```diff
@@ -0,0 +1,269 @@
+ */
+        return first.andThen(second).apply(input);
+    }
+
+    /*
+    Zadanie 9
+Napisz metodę calculate(int a, int b, BiFunction<Integer, Integer, Integer> operation), która:
+
+Wykonuje operację zdefiniowaną w BiFunction na dwóch liczbach całkowitych.
+W main przetestuj z lambdami dodającymi, mnożącymi i odejmującymi.
+     */
+    public static int calculate(int a, int b, BiFunction<Integer, Integer, Integer> operation) {
+        return operation.apply(a, b);
+    }
+
+    /*
+    Zadanie 10
+Napisz metodę countIf(int[] array, Predicate<Integer> predicate), która:
+
+Zlicza liczbę elementów w tablicy spełniających predicate.
+W main przetestuj licząc liczby >10 i liczby podzielne przez 3 w przykładowej tablicy.
+     */
+    public static long countIf(int[] array, Predicate<Integer> predicate) {
+        return Arrays.stream(array).filter(predicate::test).count();
+    }
+
+    /*
+    Zadanie 11
+Napisz metodę getLengthOrZero(Optional<String> opt), która:
+
+Jeśli opt ma wartość, zwraca długość tego łańcucha (użyj map i lambdy), w przeciwnym razie zwraca 0.
+W main przetestuj dla Optional.of("lambda") i Optional.empty().
+     */
+    public static int getLengthOrZero(Optional<String> opt) {
+        String s = opt.orElseGet(() -> "");
+        return s.length();
+    }
+
+    /*
+    Zadanie 12
+Napisz metodę combineOptionals(Optional<String> opt1, Optional<String> opt2, BiFunction<String,String,String> combiner), która:
+
+Jeśli obie wartości istnieją, łączy je używając combiner.
+Jeśli którejś brakuje, zwraca Optional.empty().
+W main przetestuj, łącząc dwie nazwy np. "Hello" i "World".
+     */
+    public static Optional<String> combineOptionals(Optional<String> opt1, Optional<String> opt2, BiFunction<String, String, String> combiner) {
+        if (opt1.isEmpty() || opt2.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(combiner.apply(opt1.get(), opt2.get()));
+        }
+    }
+
+    /*
+    Zadanie 13
+Napisz metodę flatMapToInt(Optional<String> opt), która:
+
+Jeśli wartość istnieje, próbuje sparsować ją do Integer używając flatMap i lambdy.
+Zwraca Optional<Integer> z wynikiem lub pusty Optional jeśli parse się nie uda.
+W main przetestuj dla "123", "abc" i pustego Optional.
+     */
+    public static Optional<Integer> flatMapToInt(Optional<String> opt) {
+        Optional<Integer> optional = opt.flatMap(s -> Optional.of(Integer.valueOf(s)));
+        return optional;
+    }
+
+    /*
+    Zadanie 14
+Napisz metodę transformAndSum(int[] array, Function<Integer,Integer> transformer), która:
+
+Stosuje transformer do każdego elementu tablicy i sumuje wyniki.
+Zwraca sumę.
+W main przetestuj z lambdą podwajającą liczby i lambdą zwracającą kwadrat liczby.
+     */
+    public static int transformAndSum(int[] array, Function<Integer, Integer> transformer) {
+        return Arrays.stream(array).map(transformer::apply).reduce(0, Integer::sum);
```

Comment:
Kod jest poprawny, ale metoda Arrays.stream(int[]) tworzy strumień typu IntStream, który operuje na prymitywnych int. Natomiast Function<Integer, Integer> operuje na obiektowym Integer, co wymusza dodatkowe boxowanie i deboxowanie elementów w strumieniu. Można to zoptymalizować, zamieniając int[] na Integer[] lub używając IntStream.map.
```

return Arrays.stream(array).map(x -> transformer.apply(x)).sum();

return Arrays.stream(array).map(transformer::apply).reduce(0, Integer::sum);
```